### PR TITLE
Features/allocation transfer 333

### PR DESCRIPTION
**Pull request by andrenerd**
Add method to transfer allocation from one owner to another.
The flow:
- update allocation owner's write pool to transfer non-used tokens from the related allocation pools
- update and save several allocation fields

PS. related to allocation challenge pools are owner-independent and remain unchanged

### CODE REVIEW

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 1]
```diff
@@ -490,12 +498,7 @@ func (sc *StorageSmartContract) saveUpdatedAllocation(all *StorageNodes,
 		}
 	}
 
-	// save allocation
-	_, err = balances.InsertTrieNode(alloc.GetKey(sc.ID), alloc)
-	if err != nil {
-		return
-	}
-
+	err = sc.saveAllocation(alloc, balances)
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go:501]: Why not 
```go
return sc.saveAllocation(alloc, balances)
```
?

File: code/go/0chain.net/smartcontract/storagesc/writepool.go [hunk 2]
```diff
@@ -168,6 +168,7 @@ func (wp *writePool) moveToStake(sscKey, allocID, blobID string, zcnPool tokenpo
 	return nil
 }
 */
+
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/storagesc/writepool.go:171]: If you want to refactor here, remove the commented out code.

File: code/go/0chain.net/smartcontract/storagesc/writepool.go [hunk 3]
```diff
@@ -333,6 +334,71 @@ func (ssc *StorageSmartContract) createWritePool(t *transaction.Transaction,
 	return
 }
 
+func (ssc *StorageSmartContract) transferWritePool(t *transaction.Transaction,
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/storagesc/writepool.go:337]: Needs a unit test.

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 4]
```diff
@@ -916,6 +917,115 @@ func (sc *StorageSmartContract) updateAllocationRequest(
...
+	}
+
+	return string(alloc.Encode()), nil
+}
+
+// transfer allocation and related stuff (such as, write allocation pool) to new owner
+func (sc *StorageSmartContract) transferAllocationRequest(
+	t *transaction.Transaction, input []byte,
+	balances chainstate.StateContextI,
+) (resp string, err error) {
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go:973]: Needs a unit test.

**Review by Sriep — COMMENTED** [review]
Will review again when the unit tests are done.

**Review by Sriep — CHANGES_REQUESTED** [review]
Needs unit tests.

**Comment by guruhubb** [comment]
@andrenerd @Sriep check my related comments 
https://github.com/0chain/zboxcli/issues/29#issuecomment-860120687

**Comment by Sriep** [comment]
Replaced with curator transfer https://github.com/0chain/0chain/pull/366

**Pull request closed without merging**